esp-hal/src/lib.rs

```diff
@@ -572,6 +572,115 @@ mod critical_section_impl {
     }
 }
 
+// The state of a re-entrant lock
+pub(crate) struct LockState {
+    #[cfg(multi_core)]
+    core: portable_atomic::AtomicU8,
+}
+
+impl LockState {
+    pub const fn new() -> Self {
+        Self {
+            #[cfg(multi_core)]
+            core: portable_atomic::AtomicU8::new(0),
+        }
+    }
+
+    #[cfg(multi_core)]
+    fn current_core() -> u8 {
+        // Note: 0 means unlocked.
+        match get_core() {
+            Cpu::ProCpu => 1,
+            Cpu::AppCpu => 2,
+        }
+    }
+}
+
+// This is preferred over critical-section as this allows you to have multiple
+// locks active at the same time rather than using the global mutex that is
+// critical-section.
+#[allow(unused_variables)]
+pub(crate) fn lock<T>(state: &LockState, f: impl FnOnce() -> T) -> T {
+    // In regards to disabling interrupts, we only need to disable
+    // the interrupts that may only eb calling this function.
+
+    #[cfg(not(multi_core))]
+    {
+        // Disabling interrupts is enough on single core chips to ensure mutual
+        // exclusion.
+
+        #[cfg(riscv)]
+        return riscv::interrupt::free(f);
+        #[cfg(xtensa)]
+        return xtensa_lx::interrupt::free(|_| f());
+    }
+
+    #[cfg(multi_core)]
+    {
+        use portable_atomic::Ordering;
+
+        let current_core = LockState::current_core();
+
+        // Check for re-entry
+        {
+            // Used Relaxed ordering since an Acquire has already been performed for the
+            // re-entrant case that is being checked here.
+            let locked_core = state.core.load(Ordering::Relaxed);
+            if locked_core == current_core {
+                // Note: Since this is re-entrant, interrupts have already been disabled.
+                return f();
+            }
+        }
+
+        let mut f = Some(f);
```

Comment:
I'm not entirely in love with this Some. You could return the function in the failure case, and avoid it and the interrupt. I'm not sure how it would be optimised, but it's a slightly cleaner solution.

Reply:
Same. Returning the function seems to work. It's a glorified pointer at the end of the day.

Reply:
I have some doubts in case the closure captures a lot, but hopefully LLVM is smort.